### DISCUSSION TITLE
Handle device drivers

### DISCUSSION THREAD

**Discussion by danirabbit** [elementary/switchboard-plug-about#170] [Ideas]
I've regularly seen feedback from users that seems to indicate the location of device drivers in AppCenter isn't very clear to them. Presenting drivers as updates seemed clever, but it may not be obvious, especially in the case where no drivers are available. Another thing to consider is that, in some cases, installing drivers that aren't included in the kernel can cause regressions. So it's not always really an update.

It might be clearer, especially now that Firmware will be handled here, if we also handled device drivers in the system plug. This may also make it easier to present things like Nvidia which has multiple driver options and isn't a clear linear upgrade path as in https://github.com/elementary/appcenter/issues/247

It seems also that there is a desire from users to have these more tucked away and not see them every time they run updates as in https://github.com/elementary/appcenter/issues/593 and https://github.com/elementary/appcenter/issues/285

**Comment by cassidyjames** [comment]
Yeah it _might_ make sense to handle drivers separately from OS updates and apps… a disadvantage I can see is that if there is an actual driver update that fixes a security or performance issue, it would be nice for it to be included when the user just wants to update all the things in AppCenter. But I agree it can get hairy even knowing what is a safe/recommended update versus a different version that might cause issues.

A disadvantage implementation-wise might be that it means the Switchboard Plug needs to be a PackageKit client as well as AppCenter _edit: or are drivers still an Ubuntu-specific driver utility thing?_, and we still need to figure out what is responsible for checking for updates, notifying of updates, etc. just like we will need to with firmware. We should also consider the Pop!_OS use case here to make sure they have a plan to either maintain driver support in their fork, or to handle it in their settings app.

But experience-wise I think it could make sense to have drivers—especially the initial install—be handled more like firmware in System Settings. I would be keen to hear others' thoughts as well, though.

**Reply by meisenzahl** [reply]
We could add a daemon (`io.elementary.system-daemon`). This daemon could provide the logic for installing drivers and, in the future, perhaps also for system upgrades 🤔 This daemon could also send notifications if firmware updates are available.

I'd be happy to create a prototype for it.

**Reply by danirabbit** [reply]
Re: having multiple packagekit clients, I think we had also discussed having OS updates be handled in the system plug. It might be the future of AppCenter to be flatpak only and not have packagekit there at all.

It looks like @donadigo had done some research into handling driver information in AppStream which removes Ubuntu-specific code, but it still depends on PackageKit https://github.com/elementary/appcenter/pull/243

**Reply by danirabbit** [reply]
> We could add a daemon (`io.elementary.system-daemon`). This daemon could provide the logic for installing drivers and, in the future, perhaps also for system upgrades  This daemon could also send notifications if firmware updates are available.
> 
> I'd be happy to create a prototype for it.

I don't know that we need another daemon. Packagekit operations are pretty heavy. It would probably be better to have AppCenter act as the daemon still since it already is checking for updates etc. But either way, I think it's too early to start writing code yet :)
